v3_async_wip/tests/test_iothub_session.py

```diff
@@ -0,0 +1,1289 @@
+FAKE_HOSTNAME = "fake.hostname"
+FAKE_GATEWAY_HOSTNAME = "fake.gateway.hostname"
+FAKE_URI = "fake/resource/location"
+FAKE_SHARED_ACCESS_KEY = "Zm9vYmFy"
+FAKE_SIGNATURE = "ajsc8nLKacIjGsYyB4iYDFCZaRMmmDrUuY5lncYDYPI="
+
+# TODO: aenter/aexit tests
+
+# ~~~~~ Helpers ~~~~~~
+
+
+def sastoken_generator_fn():
+    return "SharedAccessSignature sr={resource}&sig={signature}&se={expiry}".format(
+        resource=FAKE_URI, signature=FAKE_SIGNATURE, expiry=str(int(time.time()) + 3600)
+    )
+
+
+def get_expected_uri(hostname, device_id, module_id):
+    if module_id:
+        return "{hostname}/devices/{device_id}/modules/{module_id}".format(
+            hostname=hostname, device_id=device_id, module_id=module_id
+        )
+    else:
+        return "{hostname}/devices/{device_id}".format(hostname=hostname, device_id=device_id)
+
+
+# ~~~~~ Fixtures ~~~~~~
+
+# Mock out the underlying client in order to not do network operations
+@pytest.fixture(autouse=True)
+def mock_mqtt_iothub_client(mocker):
+    return mocker.patch.object(mqtt, "IoTHubMQTTClient", spec=mqtt.IoTHubMQTTClient).return_value
+
+
+# @pytest.fixture(autouse=True)
+# def mock_http_iothub_client(mocker):
+#     return mocker.patch.object(http, "IoTHubHTTPClient", spec=http.IoTHubHTTPClient).return_value
+
+
+@pytest.fixture(autouse=True)
+def mock_sastoken_provider(mocker):
+    return mocker.patch.object(st, "SasTokenProvider", spec=st.SasTokenProvider).return_value
+
+
+@pytest.fixture
+def custom_ssl_context():
+    # NOTE: It doesn't matter how the SSLContext is configured for the tests that use this fixture,
+    # so it isn't configured at all.
+    return ssl.SSLContext()
+
+
+@pytest.fixture(params=["Default SSLContext", "Custom SSLContext"])
+def optional_ssl_context(request, custom_ssl_context):
+    """Sometimes tests need to show something works with or without an SSLContext"""
+    if request.param == "Custom SSLContext":
+        return custom_ssl_context
+    else:
+        return None
+
+
+@pytest.fixture
+def session(custom_ssl_context):
+    """Use a device configuration and custom SSL auth for simplicity"""
+    return IoTHubSession(
+        hostname=FAKE_HOSTNAME, device_id=FAKE_DEVICE_ID, ssl_context=custom_ssl_context
+    )
+
+
+# ~~~~~ Parametrizations ~~~~~
+# Define parametrizations that will be used across multiple test suites, and that may eventually
+# need to be changed everywhere, e.g. new auth scheme added.
+# Note that some parametrizations are also defined within the scope of a single test suite if that
+# is the only unit they are relevant to.
+
+
+# Parameters for arguments to the __init__ or factory methods. Represent different types of
+# authentication. Use this parametrization whenever possible on .create() tests.
+# NOTE: Do NOT combine this with the SSL fixtures above. This parametrization contains
+# ssl contexts where necessary
+create_auth_params = [
+    # Provide args in form 'shared_access_key, sastoken_fn, ssl_context'
```

Comment:
># Provide args in form 'shared_access_key, sastoken_fn, ssl_context' [](http://example.com/codeflow?start=4&length=68)

recommend putting 'shared_access_key, sastoken_fn, ssl_context' into a string variable to use across test definitions,  Also "kwarg_name, kwarg_value"

Reply:
I'm not quite clear what you mean here - this list of params is here to be used across test definitions. Could you clarify?